msal/authority.py

```diff
@@ -65,12 +66,46 @@ def __init__(
             This parameter only controls whether an instance discovery will be
             performed.
         """
+        self._http_client = http_client
+        tenant_discovery_endpoint = self._initialize_entra_authority(
+            authority_url, validate_authority, instance_discovery,
+            ) if not oidc_authority_url else self._initialize_oidc_authority(
+            oidc_authority_url)
+        try:
+            openid_config = tenant_discovery(
+                tenant_discovery_endpoint,
+                self._http_client)
+        except ValueError:
+            raise ValueError(
+                "Unable to get authority configuration for {}. "
+                "Authority would typically be in a format of "
+                "https://login.microsoftonline.com/your_tenant "
+                "or https://tenant_name.ciamlogin.com "
+                "or https://tenant_name.b2clogin.com/tenant.onmicrosoft.com/policy. "
+                "Also please double check your tenant name or GUID is correct.".format(
+                authority_url))
+        logger.debug("openid_config = %s", openid_config)
+        self.authorization_endpoint = openid_config['authorization_endpoint']
+        self.token_endpoint = openid_config['token_endpoint']
+        self.device_authorization_endpoint = openid_config.get('device_authorization_endpoint')
+        _, _, self.tenant = canonicalize(self.token_endpoint)  # Usually a GUID
+
+    def _initialize_oidc_authority(self, oidc_authority_url):
+        authority, self.instance, tenant = canonicalize(oidc_authority_url)
+        self.is_adfs = tenant.lower() == 'adfs'  # As a convention
+        self._is_b2c = True  # Not exactly true, but
```

Comment:
seems iffy :). Maybe rename the flag ?

Reply:
Yeah, I thought about it. There are two approaches here.

1. Implement a cleanest `class OidcAuthority` which contains no `is_adfs`, `_is_b2c`, `_is_known_to_developer` at all. But then it would not be a drop-in replacement for the old `class Authority` (which becomes actually an EntraAuthority then). For a dozen places in `application.py` that currently reference the authority object, they would all need one extra condition in their already complicated `if` clause, such as:

```python
if isinstance(authority, OidcAuthority):
    do_oidc_thing()
else:
    assert isinstance(authority, Authority)  # It is the Entra authority
    if authority.is_adfs or authority._is_b2c:
        do_adfs_or_b2c_thing()  # Basically bypass broker code path
    elif authority._is_known_to_developer:
        skip_instance_discovery()
```

2. Bite the bullet, set `_is_b2c` and `_is_known_to_developer` as True, so that most of the existing logic in `application.py` needs no change.

After some brief attempt on `#1`, this PR is now implemented with `#2` approach. Last but not least, this is an implementation detail, so, it is a 2-way decision which we can always revisit in the future.